orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/providers/aws/AmazonMachineImageTagger.java

```diff
@@ -0,0 +1,150 @@
+
+  @Autowired
+  OortService oortService;
+
+  @Autowired
+  ObjectMapper objectMapper;
+
+  @Value("${default.bake.account:default}")
+  String defaultBakeAccount;
+
+  @Override
+  public MachineImageTagger.OperationContext getOperationContext(Stage stage) {
+    StageData stageData = (StageData) stage.mapTo(StageData.class);
+
+    MatchedImage matchedImage = findImage(stageData.amiName);
+    if (stageData.regions == null || stageData.regions.isEmpty()) {
+      stageData.regions = matchedImage.amis.keySet();
+    }
+
+    MachineImage targetMachineImage = new MachineImage(
+      stageData.amiName,
+      defaultBakeAccount,
+      stageData.regions,
+      stageData.tags
+    );
+
+    List<Map<String, Map>> operations = new ArrayList<>();
+
+    // Update the tags on the image in the `defaultBakeAccount`
+    operations.add(
+      ImmutableMap.<String, Map>builder()
+        .put(OPERATION, ImmutableMap.builder()
+          .put("amiName", targetMachineImage.imageName)
+          .put("tags", targetMachineImage.tags)
+          .put("regions", targetMachineImage.regions)
+          .put("credentials", targetMachineImage.account)
+          .build()
+        ).build()
+    );
+
+    // Re-share the image in all other accounts (will result in tags being updated)
+    matchedImage.accounts.stream()
+      .filter(account -> !account.equalsIgnoreCase(defaultBakeAccount))
+      .forEach(account -> {
+        stageData.regions.forEach(region ->
+          operations.add(
+            ImmutableMap.<String, Map>builder()
+              .put(ALLOW_LAUNCH_OPERATION, ImmutableMap.builder()
+                .put("account", account)
+                .put("credentials", defaultBakeAccount)
+                .put("region", region)
+                .put("amiName", stageData.amiName)
+                .build()
+              ).build()
+          )
+        );
+      });
+
+    Map<String, Object> extraOutput = objectMapper.convertValue(stageData, Map.class);
+    extraOutput.put("target", targetMachineImage);
+    return new MachineImageTagger.OperationContext(operations, extraOutput);
+  }
+
+  /**
+   * Return true iff the tags on the current machine image match the desired.
+   */
+  @Override
+  public boolean isMachineImageTagged(MachineImageTagger.MachineImage targetMachineImage, Stage stage) {
+    MatchedImage matchedImage = findImage(targetMachineImage.imageName);
+
+    AtomicBoolean isUpserted = new AtomicBoolean(true);
+    targetMachineImage.regions.forEach(region -> {
+        List<String> imagesForRegion = matchedImage.amis.get(region);
+        imagesForRegion.forEach(image -> {
+          Map<String, String> allImageTags = matchedImage.tagsByImageId.get(image);
+          targetMachineImage.tags.entrySet().forEach(entry -> {
+            // assert tag equality
+            isUpserted.set(isUpserted.get() && allImageTags.get(entry.getKey()).equals(entry.getValue()));
```

Comment:
would it make sense to have a set of tags that we won't update such as 'appversion' and 'buildhost'?